### PR TITLE
Add CL_MEM_ALLOC_BUFFER_LOCATION_INTEL enum

### PR DESCRIPTION
**Pull request by sherry-yuan**
This allows the user to specify which device memory to allocate to.

CC: @mkinsner @bashbaug @GarveyJoe

### CODE REVIEW

**Comment by mkinsner** [comment]
LGTM

File: extensions/cl_intel_unified_shared_memory.asciidoc
Comment (EwanC) [inline extensions/cl_intel_unified_shared_memory.asciidoc:726]: Does the `cl_intel_mem_alloc_buffer_location` type need defined, or is it defined somewhere already? Think `intel` part should also be moved so it's a suffix of the type.

Comment (sherry-yuan) [reply extensions/cl_intel_unified_shared_memory.asciidoc:726]: `cl_intel_mem_alloc_buffer_location` is defined in `cl_intel_mem_alloc_buffer_location.asciidoc`, I kept intel to be at the beginning because all other extensions eg. `cl_intel_mem_channel_property.asciidoc` uses `intel` before type (specifically it uses `cl_intel_mem_channel_property`).

Comment (EwanC) [reply extensions/cl_intel_unified_shared_memory.asciidoc:726]: The naming convention for the extension name itself is all good, but the convention for types is different and is to have the suffix at the end. E.g `cl_mem_properties_intel`, `cl_mem_alloc_flags_intel`.  So i'd expect to see this similarly defined somewhere like `typedef foo cl_mem_alloc_buffer_location _intel`

Comment (sherry-yuan) [reply extensions/cl_intel_unified_shared_memory.asciidoc]: Thank you, I have changed it to align with the types in `cl_intel_mem_alloc_buffer_location.asciidoc`

File: xml/cl.xml
Comment (bashbaug) [inline xml/cl.xml:6330]: This enum needs to be defined in its own block specifically for this new extension vs. adding it ot the USM block.  It's very possible that an implementation will support USM but not this enum.

The XML block specifically for this new extension could look something like:

```xml
        <extension name="cl_mem_alloc_buffer_location_intel" supported="opencl">
            <require>
                <type name="CL/cl.h"/>
            </require>
            <require comment="cl_mem_properties_intel">
                <enum name="CL_MEM_ALLOC_BUFFER_LOCATION_INTEL"/>
            </require>
            <require comment="cl_mem_alloc_info_intel">
                <enum name="CL_MEM_ALLOC_BUFFER_LOCATION_INTEL"/>
            </require>
        </extension>
```

Debatably there should also be sections for `cl_mem_properties` and `cl_mem_info` but this won't matter in practice.

File: extensions/cl_intel_unified_shared_memory.asciidoc
Comment (bashbaug) [inline extensions/cl_intel_unified_shared_memory.asciidoc:238]: I'm fine saying that the USM extension interacts with the new `cl_intel_mem_alloc_buffer_location` extension (see the [section](https://github.com/KhronosGroup/OpenCL-Docs/blob/master/extensions/cl_extension_template.asciidoc#interactions-with-other-extensions) of the extension template that describes how to do this), but I don't think that this enum should be listed directly in the USM spec.  Many USM implementations will not support the buffer location feature.

File: extensions/cl_intel_unified_shared_memory.asciidoc
Comment (bashbaug) [inline extensions/cl_intel_unified_shared_memory.asciidoc:729]: This description should also move out of the USM spec and into the `cl_intel_mem_alloc_buffer_location` extension spec.

**Review by bashbaug — CHANGES_REQUESTED** [review]
The content of this extension looks good but I think we need to move a few things around to keep this extension distinct from the USM extension.  Thanks!

File: xml/cl.xml
Comment (sherry-yuan) [reply xml/cl.xml]: Thank you! I have moved buffer location reference in USM to extension's own definition (both in xml and asciidoc). Specifically:

Moved the definition in xml from `cl_mem_properties_intel` and `cl_mem_alloc_info_intel` to self contained `<extension name="cl_mem_alloc_buffer_location_intel"...`

Also removed extention's reference in `cl_intel_unified_shared_memory.asciidoc` and moved table buffer location row of `cl_mem_info_intel` to `extensions/cl_intel_mem_alloc_buffer_location.asciidoc`

Comment (bashbaug) [reply xml/cl.xml]: Thanks, the XML updates look good now.

File: extensions/cl_intel_unified_shared_memory.asciidoc
Comment (bashbaug) [reply extensions/cl_intel_unified_shared_memory.asciidoc:238]: This change still needs to be reverted, and an "Interactions with Other Extensions" could be added in its place.

File: extensions/cl_intel_unified_shared_memory.asciidoc
Comment (bashbaug) [reply extensions/cl_intel_unified_shared_memory.asciidoc:728]: I see that this section has been moved to the `cl_intel_mem_alloc_buffer_location` spec but it hasn't been removed from the USM spec.

File: extensions/cl_intel_unified_shared_memory.asciidoc
Comment (sherry-yuan) [reply extensions/cl_intel_unified_shared_memory.asciidoc]: I have removed all reference of buffer_location, and moved some info under interaction section. Sorry I originally misunderstood it to be adding "interaction" to buffer_location extension doc (the interaction section is still present in buffer_location doc).

**Review by bashbaug — APPROVED** [review]
Thanks, all of my issues have been addressed, LGTM.

@mkinsner @garveyjoe @EwanC would you like to take one last look before merging?

**Review by EwanC — APPROVED** [review]
LGTM, thanks for the changes to help keep this distinct from USM

File: xml/cl.xml
Comment (sherry-yuan) [reply xml/cl.xml]: Thank you, I have changed its name and squashed the commits :)

**Review by mkinsner — APPROVED** [review]
LGTM

**Pull request merged**
